src/S3Writer.php

```diff
@@ -77,21 +77,7 @@ public function execute(string $sourcePath): void
                 );
             }
         } catch (S3Exception $e) {
-            if ($e->getStatusCode() === 403) {
-                throw new UserException("Invalid credentials or permissions.", $e->getCode(), $e);
-            }
-            if ($e->getStatusCode() === 400 || $e->getStatusCode() === 401 || $e->getStatusCode() === 404) {
-                if ($e->getPrevious() instanceof ClientException) {
-                    /** @var ClientException $previous */
-                    $previous = $e->getPrevious();
-                    if ($previous->getResponse()) {
-                        throw S3WriterException::factory($e);
-                    }
-                    throw new UserException($previous->getMessage());
-                }
-                throw new UserException($e->getMessage());
-            }
-            throw $e;
+            throw S3WriterException::fromS3Exception($e);
```

Comment:
Tyjo, sorry. Ještě ne. IMHO by neměl ten statický konstruktor vracet tu obecnou exception. Měl by mít `fromS3Exception(S3Exception $e): self`. A místo všech `return new UserException` by měl vracet `return new self` (a být `S3UserException extends UserException`. 

Ale asi to není potřeba celé přepisovat, holt jsem to asi špatně vysvětlil. Teď by mi asi stačilo to přejmenovat z `S3WriterException` (není to exception) na `S3WriterExceptionConverter`, aby to bylo obvious. Já si doteď nevšimnul, že ta třída `S3WriterException` není exception ale nějaká služba... Můžem si případně okolo toho zavolat, kdybys chtěl víc pochopit, jak jsem to myslel.

Reply:
No, tomuhle vcelku rozumím. Přístup je +- stejnej, jenom se současná classa přejmenuje a vznikne nová třída výjimky. Přišlo mi to v tuhle chvíli trochu zbytečný, když to jde přímočaře a  

- `UserException` se jinde nepoužívá, a tak nemusím mít různý typy výjimek
- nepotřebuju s výjimkou nic dalšího dělat, a `UserException` na funkcionalitu úplně stačí
- kdybych chtěl fakt řešit vlastní výjimku, tak by měla bejt vlastní třída pro každou větev v `fromS3Exception`, abych nemusel psát ručně `$message`

Na `fromS3Exception(S3Exception $e): self` mě nejvíc dráždí, že 
`: self` tam bejt nemůže. Pokud `S3Exception` obsahuje 500, nemůžu vyhazovat `instanceof UserException`.

Co kdybych to přejmenoval na `S3WriterExceptionFactory` a nechal to bejt? Dávalo by ti to takhle smysl? Hangoutovat nemusíme, stejně si musím každej tvůj komentář promyslet a odpovědět na něj mi docela trvá :-)

Reply:
Oprava, přejmenoval jsem to na `ExceptionFactory`, čekám na další bramboru, odpoledne jich bude dost na výrobu kaše :-)